src/gp/src/GPMSAOptions.C

```diff
@@ -271,6 +297,293 @@ GPMSAOptions::~GPMSAOptions()
+                                              double range_max)
+{
+  if (i >= m_uncertainScaleMin.size())
+  {
+    m_uncertainScaleMin.resize(i+1, 0);
+    m_uncertainScaleRange.resize(i+1, 1);
+  }
+  m_uncertainScaleMin[i] = range_min;
+  m_uncertainScaleRange[i] = range_max - range_min;
+}
+
+
+void
+GPMSAOptions::set_scenario_parameter_scaling(unsigned int i,
+                                             double range_min,
+                                             double range_max)
+{
+  if (i >= m_scenarioScaleMin.size())
+  {
+    m_scenarioScaleMin.resize(i+1, 0);
+    m_scenarioScaleRange.resize(i+1, 1);
+  }
+  m_scenarioScaleMin[i] = range_min;
+  m_scenarioScaleRange[i] = range_max - range_min;
+}
+
+
+template <typename V>
+void
+GPMSAOptions::set_final_scaling
+  (const std::vector<typename SharedPtr<V>::Type> & m_simulationScenarios,
+   const std::vector<typename SharedPtr<V>::Type> & m_simulationParameters,
+   const std::vector<typename SharedPtr<V>::Type> & m_simulationOutputs,
+   const std::vector<typename SharedPtr<V>::Type> & m_experimentScenarios,
+   const std::vector<typename SharedPtr<V>::Type> & m_experimentOutputs)
+{
+  if ((m_autoscaleMinMaxAll && m_autoscaleMeanVarAll) ||
+      ((m_autoscaleMinMaxAll || m_autoscaleMeanVarAll) &&
+       (!m_autoscaleMinMaxUncertain.empty() ||
+        !m_autoscaleMeanVarUncertain.empty() ||
+        !m_autoscaleMinMaxScenario.empty() ||
+        !m_autoscaleMeanVarScenario.empty())))
+    queso_error_msg("Cannot autoscale based on incompatible criteria");
+
+  unsigned int dimScenario = m_simulationScenarios.size() ?
+          m_simulationScenarios[0]->sizeGlobal() : 0;
+
+  unsigned int dimParameter = m_simulationParameters.size() ?
+          m_simulationParameters[0]->sizeGlobal() : 0;
+
+  if (m_autoscaleMinMaxAll || !m_autoscaleMinMaxScenario.empty())
+    {
+      V maxScenario(*m_simulationScenarios[0]);
+
+      V minScenario(*m_simulationScenarios[0]);
+
+      for (unsigned int i=1; i < m_simulationScenarios.size(); ++i)
+        for (unsigned int p=0; p != dimScenario; ++p)
+          {
+            minScenario[p] = std::min(minScenario[p],
+                                      (*m_simulationScenarios[i])[p]);
+            maxScenario[p] = std::max(maxScenario[p],
+                                      (*m_simulationScenarios[i])[p]);
+          }
+
+      for (unsigned int i=0; i < m_experimentScenarios.size(); ++i)
+        for (unsigned int p=0; p != dimScenario; ++p)
+          {
+            minScenario[p] = std::min(minScenario[p],
+                                      (*m_experimentScenarios[i])[p]);
+            maxScenario[p] = std::max(maxScenario[p],
+                                      (*m_experimentScenarios[i])[p]);
+          }
+
+      for (unsigned int p=0; p != dimScenario; ++p)
+        if (m_autoscaleMinMaxAll ||
+            m_autoscaleMinMaxScenario.count(p))
+          {
+            if ((m_scenarioScaleMin.size() > p) &&
```

Comment:
After scaling is finalized, it will be of size dimScenario.

Beforehand, it's possible for users to set a scaling manually for a particular index, in which case we'll have resized to hold that index but not to hold every index.